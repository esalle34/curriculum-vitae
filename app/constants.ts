enum Pages {
  CV = "cv",
  COMMON = "common",
}

export { Pages };
