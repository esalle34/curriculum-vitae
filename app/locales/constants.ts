enum Languages {
  English = "en",
  Français = "fr",
  "Español" = "es",
}

enum Countries {
  France = "FR",
  USA = "US",
  "United Kingdom" = "GB",
  "España" = "ES",
  Luxembourg = "LU",
  Suisse = "CH",
  Belgique = "BE",
}

const languagesArray = Object.entries(Languages);
const countriesArray = Object.entries(Countries);

const locales = countriesArray.flatMap(([_countryKey, country]) =>
  languagesArray.map(
    ([_languageKey, language]): Locale => `${language}-${country}`
  )
);

export type Locale = `${Languages}-${Countries}`;
const defaultLocale: Locale = `${Languages.English}-${Countries.USA}`;

interface DecodeTranslationOptions {
  className?: string;
  htmlTag: string;
}

export interface DecodeTranslation {
  options: DecodeTranslationOptions;
  translation: string;
}

export {
  Languages,
  Countries,
  defaultLocale,
  languagesArray,
  countriesArray,
  locales,
};
