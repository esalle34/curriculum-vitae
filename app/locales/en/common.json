{
  "title": "Eric Salle",
  "submit": "Submit",
  "chooseLanguage": "Language & Country",
  "close": "Close"
}
