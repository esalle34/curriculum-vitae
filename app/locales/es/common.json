{
  "title": "Eric Salle",
  "submit": "Confirmar",
  "chooseLanguage": "Lengua i Pa√≠s",
  "close": "Cerrar"
}
