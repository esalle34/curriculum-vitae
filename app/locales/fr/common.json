{
  "title": "Eric Salle",
  "submit": "Valider",
  "chooseLanguage": "Langue & Pays",
  "close": "Fermer"
}
