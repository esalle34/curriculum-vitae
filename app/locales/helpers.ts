import { Countries, DecodeTranslation, Languages, Locale } from "./constants";

const getDictionary = (language: Languages, page: string) => ({
  [language]: () =>
    import(`@/app/locales/${language}/${page}.json`).then(
      (module) => module.default
    ),
});

const getLanguageAndCountry = (locale: Locale) => {
  const [language, country] = locale.split("-");

  return { language: language as Languages, country: country as Countries };
};

const decodeTranslation = ({
  options: { htmlTag, className },
  translation,
}: DecodeTranslation) => {
  return translation
    .replace(`<0>`, `<${htmlTag} ${className ? `class="${className}"` : ""}>`)
    .replace(`</0>`, `</${htmlTag}>`);
};

export { getDictionary, getLanguageAndCountry, decodeTranslation };
