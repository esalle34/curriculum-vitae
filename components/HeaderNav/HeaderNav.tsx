"use client";
import { useRouter } from "next/navigation";
import CountrySelect from "../CountrySelect";
import LanguageSelect from "../LanguageSelect";
import Popin from "../Popin";
import { useState } from "react";
import { Countries, Languages } from "@/app/locales/constants";
import getUnicodeFlagIcon from "country-flag-icons/unicode";

interface Props {
  country: Countries;
  language: Languages;
  dict: Record<string, string>;
}

const HeaderNav = ({ country, language, dict }: Props) => {
  const router = useRouter();
  const [displayed, setDisplayed] = useState(false);
  const [newCountry, setNewCountry] = useState(country);
  const [newLanguage, setNewLanguage] = useState(language);
  return (
    <nav className="flex">
      <div className="flex w-full justify-between items-center">
        <a href="https://www.linkedin.com/in/%F0%9F%92%BB-eric-salle-431664109/">
          <h1 className="pl-4 text-gray-100 font-bold text-2xl my-2 font-instagram">
            {dict.title}
          </h1>
        </a>
        <button
          className="mr-4 my-4 text-gray-100 font-semibold border-2 border-color-white p-2 rounded hover:bg-white hover:text-sky-800"
          onClick={() => setDisplayed(true)}
        >
          {language.toUpperCase()} - {getUnicodeFlagIcon(country)}
        </button>
      </div>
      <Popin
        displayed={displayed}
        setDisplayed={setDisplayed}
        title={dict.chooseLanguage}
        closeText={dict.close}
      >
        <CountrySelect country={country} setNewCountry={setNewCountry} />
        <LanguageSelect language={language} setNewLanguage={setNewLanguage} />
        <button
          className="self-end rounded-lg text-sky-800 border-sky-800 hover:bg-sky-800 hover:text-white border-2 p-2"
          onClick={() => {
            router.push(`${newLanguage}-${newCountry}`);
            setDisplayed(false);
          }}
        >
          {dict.submit}
        </button>
      </Popin>
    </nav>
  );
};

export default HeaderNav;
