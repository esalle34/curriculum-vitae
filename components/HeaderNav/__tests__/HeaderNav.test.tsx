import { Countries, Languages } from "@/app/locales/constants";
import * as dict from "@/app/locales/fr/common.json";
import HeaderNav from "..";
import { fireEvent, render } from "@testing-library/react";

jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      prefetch: () => null,
      push: () => null,
    };
  },
}));

describe("HeaderNav", () => {
  it("should match FR - France snapshot", () => {
    const { container } = render(
      <HeaderNav
        country={Countries.France}
        language={Languages.Français}
        dict={dict}
      />
    );
    expect(container).toMatchSnapshot();
  });
  it("should match EN - USA snapshot", () => {
    const { container } = render(
      <HeaderNav
        country={Countries.USA}
        language={Languages.English}
        dict={dict}
      />
    );
    expect(container).toMatchSnapshot();
  });
  it("should match ES - Spain snapshot", () => {
    const { container } = render(
      <HeaderNav
        country={Countries.España}
        language={Languages.Español}
        dict={dict}
      />
    );
    expect(container).toMatchSnapshot();
  });
  it("should open and close Popin", () => {
    const { queryAllByRole, getByTestId } = render(
      <HeaderNav
        country={Countries.España}
        language={Languages.Español}
        dict={dict}
      />
    );
    const buttons = queryAllByRole("button");
    expect(getByTestId("popin")).toHaveClass("hidden");
    fireEvent.click(buttons[0]);
    expect(getByTestId("popin")).not.toHaveClass("hidden");
    fireEvent.click(buttons[1]);
    expect(getByTestId("popin")).toHaveClass("hidden");
  });
});
