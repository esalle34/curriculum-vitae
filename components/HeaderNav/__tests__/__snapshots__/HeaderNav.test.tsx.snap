// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderNav should match EN - USA snapshot 1`] = `
<div>
  <nav
    class="flex"
  >
    <div
      class="flex w-full justify-between items-center"
    >
      <h1
        class="pl-4 text-gray-100 font-bold text-lg my-2"
      >
        Eric Salle
      </h1>
      <button
        class="mr-4 my-4 text-gray-100 font-semibold border-2 border-color-white p-2 rounded hover:bg-white hover:text-blue-800"
      >
        EN
         - 
        🇺🇸
      </button>
    </div>
    <div
      class="hidden"
      data-testid="popin"
    >
      <div
        class="flex flex-col gap-2 md:w-1/4 bg-slate-100 py-4 px-6 rounded z-10 border-2 border-slate-200"
      >
        <div
          class="flex justify-between items-center"
        >
          <h3
            class="py-2 text-lg font-semibold text-blue-800"
          >
            Langue & Pays
          </h3>
          <img
            alt="Fermer"
            class="cursor-pointer"
            data-nimg="1"
            decoding="async"
            height="20"
            loading="lazy"
            role="img"
            src="/images/cross.svg"
            style="color: transparent;"
            title="Fermer"
            width="20"
          />
        </div>
        <select
          class="my-2 p-2 border-2 border-slate-300 rounded"
        >
          <option
            value="FR"
          >
            France
             - 
            🇫🇷
          </option>
          <option
            selected=""
            value="US"
          >
            USA
             - 
            🇺🇸
          </option>
          <option
            value="GB"
          >
            United Kingdom
             - 
            🇬🇧
          </option>
          <option
            value="ES"
          >
            España
             - 
            🇪🇸
          </option>
          <option
            value="LU"
          >
            Luxembourg
             - 
            🇱🇺
          </option>
          <option
            value="CH"
          >
            Suisse
             - 
            🇨🇭
          </option>
          <option
            value="BE"
          >
            Belgique
             - 
            🇧🇪
          </option>
        </select>
        <select
          class="my-2 p-2 border-2 border-slate-300 rounded"
        >
          <option
            selected=""
            value="en"
          >
            English
          </option>
          <option
            value="fr"
          >
            Français
          </option>
          <option
            value="es"
          >
            Español
          </option>
        </select>
        <button
          class="self-end rounded-lg text-blue-800 border-blue-800 hover:bg-blue-800 hover:text-white border-2 p-2"
        >
          Valider
        </button>
      </div>
      <div
        class="hidden"
        data-testid="background"
      />
    </div>
  </nav>
</div>
`;

exports[`HeaderNav should match ES - Spain snapshot 1`] = `
<div>
  <nav
    class="flex"
  >
    <div
      class="flex w-full justify-between items-center"
    >
      <h1
        class="pl-4 text-gray-100 font-bold text-lg my-2"
      >
        Eric Salle
      </h1>
      <button
        class="mr-4 my-4 text-gray-100 font-semibold border-2 border-color-white p-2 rounded hover:bg-white hover:text-blue-800"
      >
        ES
         - 
        🇪🇸
      </button>
    </div>
    <div
      class="hidden"
      data-testid="popin"
    >
      <div
        class="flex flex-col gap-2 md:w-1/4 bg-slate-100 py-4 px-6 rounded z-10 border-2 border-slate-200"
      >
        <div
          class="flex justify-between items-center"
        >
          <h3
            class="py-2 text-lg font-semibold text-blue-800"
          >
            Langue & Pays
          </h3>
          <img
            alt="Fermer"
            class="cursor-pointer"
            data-nimg="1"
            decoding="async"
            height="20"
            loading="lazy"
            role="img"
            src="/images/cross.svg"
            style="color: transparent;"
            title="Fermer"
            width="20"
          />
        </div>
        <select
          class="my-2 p-2 border-2 border-slate-300 rounded"
        >
          <option
            value="FR"
          >
            France
             - 
            🇫🇷
          </option>
          <option
            value="US"
          >
            USA
             - 
            🇺🇸
          </option>
          <option
            value="GB"
          >
            United Kingdom
             - 
            🇬🇧
          </option>
          <option
            selected=""
            value="ES"
          >
            España
             - 
            🇪🇸
          </option>
          <option
            value="LU"
          >
            Luxembourg
             - 
            🇱🇺
          </option>
          <option
            value="CH"
          >
            Suisse
             - 
            🇨🇭
          </option>
          <option
            value="BE"
          >
            Belgique
             - 
            🇧🇪
          </option>
        </select>
        <select
          class="my-2 p-2 border-2 border-slate-300 rounded"
        >
          <option
            value="en"
          >
            English
          </option>
          <option
            value="fr"
          >
            Français
          </option>
          <option
            selected=""
            value="es"
          >
            Español
          </option>
        </select>
        <button
          class="self-end rounded-lg text-blue-800 border-blue-800 hover:bg-blue-800 hover:text-white border-2 p-2"
        >
          Valider
        </button>
      </div>
      <div
        class="hidden"
        data-testid="background"
      />
    </div>
  </nav>
</div>
`;

exports[`HeaderNav should match FR - France snapshot 1`] = `
<div>
  <nav
    class="flex"
  >
    <div
      class="flex w-full justify-between items-center"
    >
      <h1
        class="pl-4 text-gray-100 font-bold text-lg my-2"
      >
        Eric Salle
      </h1>
      <button
        class="mr-4 my-4 text-gray-100 font-semibold border-2 border-color-white p-2 rounded hover:bg-white hover:text-blue-800"
      >
        FR
         - 
        🇫🇷
      </button>
    </div>
    <div
      class="hidden"
      data-testid="popin"
    >
      <div
        class="flex flex-col gap-2 md:w-1/4 bg-slate-100 py-4 px-6 rounded z-10 border-2 border-slate-200"
      >
        <div
          class="flex justify-between items-center"
        >
          <h3
            class="py-2 text-lg font-semibold text-blue-800"
          >
            Langue & Pays
          </h3>
          <img
            alt="Fermer"
            class="cursor-pointer"
            data-nimg="1"
            decoding="async"
            height="20"
            loading="lazy"
            role="img"
            src="/images/cross.svg"
            style="color: transparent;"
            title="Fermer"
            width="20"
          />
        </div>
        <select
          class="my-2 p-2 border-2 border-slate-300 rounded"
        >
          <option
            selected=""
            value="FR"
          >
            France
             - 
            🇫🇷
          </option>
          <option
            value="US"
          >
            USA
             - 
            🇺🇸
          </option>
          <option
            value="GB"
          >
            United Kingdom
             - 
            🇬🇧
          </option>
          <option
            value="ES"
          >
            España
             - 
            🇪🇸
          </option>
          <option
            value="LU"
          >
            Luxembourg
             - 
            🇱🇺
          </option>
          <option
            value="CH"
          >
            Suisse
             - 
            🇨🇭
          </option>
          <option
            value="BE"
          >
            Belgique
             - 
            🇧🇪
          </option>
        </select>
        <select
          class="my-2 p-2 border-2 border-slate-300 rounded"
        >
          <option
            value="en"
          >
            English
          </option>
          <option
            selected=""
            value="fr"
          >
            Français
          </option>
          <option
            value="es"
          >
            Español
          </option>
        </select>
        <button
          class="self-end rounded-lg text-blue-800 border-blue-800 hover:bg-blue-800 hover:text-white border-2 p-2"
        >
          Valider
        </button>
      </div>
      <div
        class="hidden"
        data-testid="background"
      />
    </div>
  </nav>
</div>
`;
